File: LinkedList/LinkedList.rb
# Classic linked list implementation with basic class methods

class Element
  attr_accessor :next, :value

  def initialize(value)
    @next = nil
    @value = value
  end
end

class LinkedList
  attr_accessor :name

  def initialize
    @head = nil
    @tail = nil
  end

  # Puts an element at the beginning of the LinkedList (ptq)
  def unshift(element)
    if @head.nil?
      @head = element
      @tail = element
    else
      element.next = @head
      @head = element
    end
    return self
  end

  # Puts an element at the end of the LinkedList (pbq)
  def push(element)
    if @head.nil?
      @head = element
      @tail = element
    else
      @tail.next = element
      @tail = element
    end
    return self
  end

  # Removes first element from the beginning of the LinkedList and returns it (rtq)
  def shift
    return nil if @head.nil?
    element = @head
    @head = @head.next
    return element
  end

  # Removes the last element from the end of the LL and returns it
  def pop
    return nil if @head.nil?
    element = @tail
    @tail = self.find_by_next(element)
    @tail.next = nil unless @tail.nil?
    return element
  end

  # Returns the last element whose next points to the argument element
  def find_by_next(element)
    result = nil

    self.each do |node|
      result = node if node.next == element
    end
    return result
  end

  include Enumerable

  def each 
    return nil if @head.nil?
    element = @head

    until element.nil?
      yield element
      element = element.next
    end
  end

  # Converts LinkedLists to Arrays
  def to_a
    a = Array.new

    self.each do |element|
      a.push(element.value)
    end

    return a
  end
end

class Array
  #converts arrays to LinkedLists
  def to_l
    ll = LinkedList.new

    self.each do |element|
      l_element = Element.new(element)
      ll.push(l_element)
    end

    return ll
  end
end

# basic unit tests
def test
  a = [1,2,3,4,5,6]
  b = ["a","b",1,2,-3,54.5]
  c = [0,1,2,2,1,0]
  d = []

  def test_to_l_and_to_a(a,b,c,d)
    puts "test to_l and to_a"
    puts a == a.to_l.to_a
    puts b == b.to_l.to_a
    puts c == c.to_l.to_a
    puts d == d.to_l.to_a
  end

  def test_unshift(a,b,c,d)
    puts "test unshift"
    puts a.to_l.unshift(Element.new(10)).to_a == a.unshift(10)
    puts b.to_l.unshift(Element.new(10)).to_a == b.unshift(10)
    puts c.to_l.unshift(Element.new(10)).to_a == c.unshift(10)
    puts d.to_l.unshift(Element.new(10)).to_a == d.unshift(10)
  end

  def test_push(a,b,c,d)
    puts "test push"
    puts a.to_l.push(Element.new(10)).to_a == a.push(10)
    puts b.to_l.push(Element.new(10)).to_a == b.push(10)
    puts c.to_l.push(Element.new(10)).to_a == c.push(10)
    puts d.to_l.push(Element.new(10)).to_a == d.push(10)
  end

  def test_shift(a,b,c,d)
    puts "test shift"
    puts a.to_l.shift.value == a.shift
    puts b.to_l.shift.value == b.shift
    puts c.to_l.shift.value == c.shift
    puts d.to_l.shift.value == d.shift
  end

  def test_pop(a,b,c,d)
    puts "test pop"
    aa = a.to_l
    
    puts aa.pop.value == a.pop
    puts aa.pop.value == a.pop
    puts b.to_l.pop.value == b.pop
    puts c.to_l.pop.value == c.pop
    puts d.to_l.pop.value == d.pop
  end

  def test_find_by_next(a,b,c,d)
    puts "test find_by_next"
    aa = a.to_l
    bb = b.to_l
    cc = c.to_l

    aa.each do |element|
      puts aa.find_by_next(element.next) == element
    end

    bb.each do |element|
      puts bb.find_by_next(element.next) == element
    end

    cc.each do |element|
      puts cc.find_by_next(element.next) == element
    end
  end

  test_to_l_and_to_a(a,b,c,d)
  test_unshift(a,b,c,d)
  test_push(a,b,c,d)
  test_shift(a,b,c,d)
  test_pop(a,b,c,d)
  test_find_by_next(a,b,c,d)
end

test

File: javascript/LinkedList/LinkedList.js
// Define an Element Class
function Element(value) {
  this.next = null;
  this.value = value;
}

// Define Linked List class. Optionally takes an array of values.
function LinkedList(values) {
  var self = this;
  this.head = null;
  this.tail = null;

  if (values) {
    values.map( function(value) {
      element = new Element(value);
      self.push(element);
    })
  }
  return this;
}

// Puts an element at the beginning of the LinkedList (ptq)
LinkedList.prototype.unshift = function(element) {
  if (this.head === null) {
    this.head = element;
    this.tail = element;
  }
  else {
    element.next = this.head;
    this.head = element;
  }
  return this;
}


// Puts an element at the end of the LinkedList (pbq)
LinkedList.prototype.push = function(element) {
  if (this.head === null) {
    this.head = element;
    this.tail = element;
  }
  else {
    this.tail.next = element;
    this.tail = element;
  }
  return this;
}

// Removes first element from the beginning of the LinkedList and returns it (rtq)
LinkedList.prototype.shift = function() {
  if (this.head === null) return null;
  element = this.head;
  this.head = this.head.next;
  return element;
}

// Removes the last element from the end of the LL and returns it
LinkedList.prototype.pop = function() {
  if (this.head === null) return null;
  element = this.tail;
  this.tail = this.findByNext(element);
  if (this.tail != null) this.tail.next = null;
  return element;
}

// Returns the last element whose next points to the argument element
LinkedList.prototype.findByNext = function(element) {
  var result = null;
  var node = this.head

  while (node !== null) {
    if (node.next === element) result = node;
    node = node.next;
  }
  return result;
}

// Returns each element of a linked list sequentially
// LinkedList.prototype.each = function*() {
//   if (this.head === null) return null;
//   var element = this.head;

//   while (element !== null) {
//     yield element;
//     element = element.next;
//   }
// }

// Displays the contents of a LinkedList as an array
LinkedList.prototype.display = function() {
  if (this.head === null) return [];
  var array = new Array;
  var element = this.head;

  while (element !== null) {
    array.push(element.value);
    element = element.next;
  }

  return array;
}

// Tests array equality
function arraysEqual(a, b) {
  if (a === b) return true;
  if (a == null || b == null) return false;
  if (a.length != b.length) return false;

  for (var i = 0; i < a.length; ++i) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}


function TestData() {
  this.a = new LinkedList([1,2,3,4,5,6]);
  this.b = new LinkedList(["a",1,2,-3,54.5]);
  this.c = new LinkedList([0,1,2,2,1,0]);
  this.d = new LinkedList([]);
}

// TestData.prototype.each = function() {

// }

TestData.prototype.testDisplay = function() {
  console.log("Display Tests")
  console.log(this.a.display())
  console.log(arraysEqual(this.a.display(),[1,2,3,4,5,6]));
  console.log(arraysEqual(this.b.display(),["a",1,2,-3,54.5]));
  console.log(arraysEqual(this.c.display(),[0,1,2,2,1,0]));
  console.log(arraysEqual(this.d.display(),[]));
}

TestData.prototype.testUnshift = function() {
  console.log("Unshift Tests")
  console.log(arraysEqual(this.a.unshift(new Element(1)).display(),[1,1,2,3,4,5,6]));
  console.log(arraysEqual(this.b.unshift(new Element(2)).display(),[2,"a",1,2,-3,54.5]));
  console.log(arraysEqual(this.c.unshift(new Element(3)).display(),[3,0,1,2,2,1,0]));
  console.log(arraysEqual(this.d.unshift(new Element(1)).display(),[1]));
  console.log(arraysEqual(this.d.unshift(new Element(2)).display(),[2,1]));
}

TestData.prototype.testPush = function() {
  console.log("Push Tests")
  console.log(arraysEqual(this.a.push(new Element(1)).display(),[1,2,3,4,5,6,1]));
  console.log(arraysEqual(this.b.push(new Element(2)).display(),["a",1,2,-3,54.5,2]));
  console.log(arraysEqual(this.c.push(new Element(3)).display(),[0,1,2,2,1,0,3]));
  console.log(arraysEqual(this.d.push(new Element(1)).display(),[1]));
  console.log(arraysEqual(this.d.push(new Element(2)).display(),[1,2]));
}

TestData.prototype.testShift = function() {
  console.log("Push Tests")
  console.log(this.a.shift().value === 1);
  console.log(this.b.shift().value === "a");
  console.log(this.c.shift().value === 0);
  console.log(this.d.shift() === null);
}

TestData.prototype.testfindByNext = function() {
  console.log("findByNext Tests")
  var element = this.a.head
  while (element !== null) {
    console.log(this.a.findByNext(element.next) === element);
    element = element.next;
  }

  var element = this.b.head
  while (element !== null) {
    console.log(this.b.findByNext(element.next) === element);
    element = element.next;
  }

  var element = this.c.head
  while (element !== null) {
    console.log(this.c.findByNext(element.next) === element);
    element = element.next;
  }

  var element = this.d.head
  while (element !== null) {
    console.log(this.d.findByNext(element.next) === element);
    element = element.next;
  }
}

TestData.prototype.testPop = function() {
  console.log("Pop Tests")
  console.log(this.a.pop().value === 6);
  console.log(this.b.pop().value === 54.5);
  console.log(this.b.pop().value === -3);
  console.log(this.c.pop().value === 0);
  console.log(this.d.pop() === null);
}

//new TestData().testEach();
new TestData().testDisplay();
new TestData().testUnshift();
new TestData().testPush();
new TestData().testShift();
new TestData().testfindByNext();
new TestData().testPop();
